File: packages/expo-audio-studio/android/src/androidTest/java/net/siteed/audiostream/integration/CompressedOnlyOutputTest.kt
package net.siteed.audiostream.integration

import android.os.Bundle
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.Assert.*
import java.io.File

/**
 * Integration test for Compressed-Only Output (Issue #244)
 * 
 * This test validates that when primary output is disabled and compressed output
 * is enabled, the compressed file information is properly returned in the result.
 */
@RunWith(AndroidJUnit4::class)
class CompressedOnlyOutputTest {

    private val context = InstrumentationRegistry.getInstrumentation().targetContext

    @Test
    fun testCompressedOnlyOutput_AAC() {
        println("🧪 Test: Compressed-Only Output with AAC")
        println("---------------------------------------")

        // Configuration with primary disabled, compressed enabled
        val config = Bundle().apply {
            putInt("sampleRate", 44100)
            putInt("channels", 1)
            putString("encoding", "pcm_16bit")
            
            val outputBundle = Bundle().apply {
                val primaryBundle = Bundle().apply {
                    putBoolean("enabled", false)
                }
                val compressedBundle = Bundle().apply {
                    putBoolean("enabled", true)
                    putString("format", "aac")
                    putInt("bitrate", 128000)
                }
                putBundle("primary", primaryBundle)
                putBundle("compressed", compressedBundle)
            }
            putBundle("output", outputBundle)
        }

        // Simulate recording result
        val result = simulateRecording(config)

        // Verify compression info is present
        val compressionBundle = result.getBundle("compression")
        assertNotNull("Compression info should not be null", compressionBundle)

        // Verify compressed file URI is provided
        val compressedFileUri = compressionBundle?.getString("compressedFileUri")
        assertNotNull("Compressed file URI should not be null", compressedFileUri)
        assertNotEquals("Compressed file URI should not be empty", "", compressedFileUri)

        // Verify format and bitrate
        assertEquals("aac", compressionBundle?.getString("format"))
        assertEquals(128000, compressionBundle?.getInt("bitrate"))

        println("✅ Compression info properly returned")
        println("✅ Compressed file URI: $compressedFileUri")
        println("✅ Format: ${compressionBundle?.getString("format")}")
        println()
    }

    @Test
    fun testCompressedOnlyOutput_Opus() {
        println("🧪 Test: Compressed-Only Output with Opus")
        println("----------------------------------------")

        val config = Bundle().apply {
            putInt("sampleRate", 48000)
            putInt("channels", 1)
            putString("encoding", "pcm_16bit")
            
            val outputBundle = Bundle().apply {
                val primaryBundle = Bundle().apply {
                    putBoolean("enabled", false)
                }
                val compressedBundle = Bundle().apply {
                    putBoolean("enabled", true)
                    putString("format", "opus")
                    putInt("bitrate", 64000)
                }
                putBundle("primary", primaryBundle)
                putBundle("compressed", compressedBundle)
            }
            putBundle("output", outputBundle)
        }

        val result = simulateRecording(config)
        val compressionBundle = result.getBundle("compression")

        assertNotNull("Compression info should not be null", compressionBundle)
        assertEquals("opus", compressionBundle?.getString("format"))
        assertEquals(64000, compressionBundle?.getInt("bitrate"))

        println("✅ Opus compression properly configured")
        println()
    }

    @Test
    fun testFileAccessibility() {
        println("🧪 Test: Verify Compressed File Accessibility")
        println("--------------------------------------------")

        val config = Bundle().apply {
            putInt("sampleRate", 44100)
            putInt("channels", 1)
            putString("encoding", "pcm_16bit")
            
            val outputBundle = Bundle().apply {
                val primaryBundle = Bundle().apply {
                    putBoolean("enabled", false)
                }
                val compressedBundle = Bundle().apply {
                    putBoolean("enabled", true)
                    putString("format", "aac")
                }
                putBundle("primary", primaryBundle)
                putBundle("compressed", compressedBundle)
            }
            putBundle("output", outputBundle)
        }

        val result = simulateRecording(config)

        // Check main result structure
        val fileUri = result.getString("fileUri", "")
        val filename = result.getString("filename", "")
        
        // Check compression structure
        val compressionBundle = result.getBundle("compression")
        val compressedUri = compressionBundle?.getString("compressedFileUri")
        val compressedSize = compressionBundle?.getLong("size", 0L) ?: 0L

        // When primary is disabled, we should have access to compressed file
        val hasAccessToCompressed = !compressedUri.isNullOrEmpty() || fileUri.isNotEmpty()
        
        assertTrue("Should have access to compressed file", hasAccessToCompressed)
        assertTrue("Compressed file should have size > 0", compressedSize > 0)

        println("✅ Compressed file is accessible")
        println("✅ File size reported: $compressedSize bytes")
        println()
    }

    @Test
    fun testBugScenario() {
        println("🧪 Test: Current Bug Scenario")
        println("-----------------------------")
        
        // This test demonstrates the current bug
        val config = Bundle().apply {
            putInt("sampleRate", 44100)
            putInt("channels", 1)
            putString("encoding", "pcm_16bit")
            
            val outputBundle = Bundle().apply {
                val primaryBundle = Bundle().apply {
                    putBoolean("enabled", false)
                }
                val compressedBundle = Bundle().apply {
                    putBoolean("enabled", true)
                    putString("format", "aac")
                    putInt("bitrate", 128000)
                }
                putBundle("primary", primaryBundle)
                putBundle("compressed", compressedBundle)
            }
            putBundle("output", outputBundle)
        }

        // Current buggy behavior
        val buggyResult = simulateBuggyRecording(config)
        
        // This is what currently happens - compression is null
        val compressionBundle = buggyResult.getBundle("compression")
        
        if (compressionBundle == null) {
            println("❌ BUG CONFIRMED: Compression info is null when primary is disabled")
            println("   This prevents users from accessing the compressed file")
        } else {
            println("✅ Bug appears to be fixed - compression info is present")
        }
        println()
    }

    /**
     * Simulates the expected correct behavior after fix
     */
    private fun simulateRecording(config: Bundle): Bundle {
        val outputConfig = config.getBundle("output")
        val primaryEnabled = outputConfig?.getBundle("primary")?.getBoolean("enabled", true) ?: true
        val compressedConfig = outputConfig?.getBundle("compressed")
        val compressedEnabled = compressedConfig?.getBoolean("enabled", false) ?: false

        return Bundle().apply {
            if (!primaryEnabled) {
                // Expected behavior after fix
                putString("fileUri", "")
                putString("filename", "stream-only")
                putLong("durationMs", 5000)
                putLong("size", 0)
                putString("mimeType", "audio/wav")
                
                // FIXED: Include compression info when compressed is enabled
                if (compressedEnabled) {
                    val compressionBundle = Bundle().apply {
                        putString("compressedFileUri", "file:///storage/emulated/0/Android/data/test/files/recording.aac")
                        putString("format", compressedConfig?.getString("format") ?: "aac")
                        putInt("bitrate", compressedConfig?.getInt("bitrate") ?: 128000)
                        putLong("size", 40000)
                        putString("mimeType", "audio/aac")
                    }
                    putBundle("compression", compressionBundle)
                }
            } else {
                // Normal behavior
                putString("fileUri", "file:///storage/emulated/0/Android/data/test/files/recording.wav")
                putString("filename", "recording.wav")
                putLong("durationMs", 5000)
                putLong("size", 240000)
                putString("mimeType", "audio/wav")
            }
        }
    }

    /**
     * Simulates the current buggy behavior
     */
    private fun simulateBuggyRecording(config: Bundle): Bundle {
        val outputConfig = config.getBundle("output")
        val primaryEnabled = outputConfig?.getBundle("primary")?.getBoolean("enabled", true) ?: true

        return Bundle().apply {
            if (!primaryEnabled) {
                // Current buggy behavior
                putString("fileUri", "")
                putString("filename", "stream-only")
                putLong("durationMs", 5000)
                putLong("size", 0)
                putString("mimeType", "audio/wav")
                // BUG: compression is null even when compressed output is enabled
                putBundle("compression", null)
            }
        }
    }
}

File: packages/audio-studio/ios/tests/integration/compressed_only_output_test.swift
#!/usr/bin/env swift

import Foundation
import AVFoundation

// Integration test for Compressed-Only Output (Issue #244)
// This tests that when primary output is disabled and compressed is enabled,
// the compressed file info is properly returned in the result

print("🧪 Compressed-Only Output Integration Test (Issue #244)")
print("=====================================================\n")

// Add the parent directory to the module search path
let srcPath = URL(fileURLWithPath: #file)
    .deletingLastPathComponent()
    .deletingLastPathComponent()
    .deletingLastPathComponent()
    .path

// Import the module
#if canImport(AudioStudio)
import AudioStudio
#endif

// Helper to load Swift files
func loadSwiftFile(_ filename: String) {
    let filePath = "\(srcPath)/\(filename).swift"
    let fileURL = URL(fileURLWithPath: filePath)
    
    do {
        let _ = try String(contentsOf: fileURL, encoding: .utf8)
        // In a real scenario, we'd compile and load this
        // For testing, we'll simulate the behavior
    } catch {
        print("Warning: Could not load \(filename).swift")
    }
}

// Test class
class CompressedOnlyOutputTest {
    var results: [(name: String, passed: Bool, message: String)] = []
    
    func runAllTests() {
        print("📋 Test Scenarios:")
        print("1. Primary disabled, Compressed enabled (AAC)")
        print("2. Primary disabled, Compressed enabled (Opus)")
        print("3. Verify compressed file URI is returned")
        print("4. Verify file size and format info")
        print("\n")
        
        testCompressedOnlyAAC()
        testCompressedOnlyOpus()
        testCompressedFileAccess()
        printResults()
    }
    
    func testCompressedOnlyAAC() {
        print("Test 1: Compressed-Only Output with AAC")
        print("---------------------------------------")
        
        // Simulate recording configuration
        let config = [
            "sampleRate": 44100,
            "channels": 1,
            "encoding": "pcm_16bit",
            "output": [
                "primary": ["enabled": false],
                "compressed": [
                    "enabled": true,
                    "format": "aac",
                    "bitrate": 128000
                ]
            ]
        ] as [String : Any]
        
        // Expected behavior: Should return compression info with file URI
        let mockResult = simulateRecording(config: config)
        
        let hasCompressionInfo = mockResult["compression"] != nil
        let compressionDict = mockResult["compression"] as? [String: Any]
        let hasCompressedUri = compressionDict?["compressedFileUri"] != nil
        let format = compressionDict?["format"] as? String
        
        let passed = hasCompressionInfo && hasCompressedUri && format == "aac"
        
        results.append((
            name: "AAC Compressed-Only",
            passed: passed,
            message: "Compression info: \(hasCompressionInfo), URI: \(hasCompressedUri), Format: \(format ?? "nil")"
        ))
        
        if passed {
            print("✅ Compression info properly returned")
            print("✅ Compressed file URI: \(compressionDict?["compressedFileUri"] ?? "nil")")
            print("✅ Format: \(format ?? "nil")")
        } else {
            print("❌ FAIL: Compression info missing or incomplete")
        }
        print()
    }
    
    func testCompressedOnlyOpus() {
        print("Test 2: Compressed-Only Output with Opus (fallback to AAC on iOS)")
        print("-----------------------------------------------------------------")
        
        let config = [
            "sampleRate": 48000,
            "channels": 1,
            "encoding": "pcm_16bit",
            "output": [
                "primary": ["enabled": false],
                "compressed": [
                    "enabled": true,
                    "format": "opus", // Should fallback to AAC on iOS
                    "bitrate": 64000
                ]
            ]
        ] as [String : Any]
        
        let mockResult = simulateRecording(config: config)
        
        let compressionDict = mockResult["compression"] as? [String: Any]
        let format = compressionDict?["format"] as? String
        let bitrate = compressionDict?["bitrate"] as? Int
        
        // On iOS, Opus should fallback to AAC
        let passed = format == "aac" && bitrate != nil
        
        results.append((
            name: "Opus→AAC Fallback",
            passed: passed,
            message: "Format: \(format ?? "nil"), Bitrate: \(bitrate ?? 0)"
        ))
        
        if passed {
            print("✅ Opus correctly fell back to AAC")
            print("✅ Bitrate preserved: \(bitrate ?? 0)")
        } else {
            print("❌ FAIL: Incorrect format or missing bitrate")
        }
        print()
    }
    
    func testCompressedFileAccess() {
        print("Test 3: Verify Compressed File Accessibility")
        print("-------------------------------------------")
        
        let config = [
            "sampleRate": 44100,
            "channels": 1,
            "encoding": "pcm_16bit",
            "output": [
                "primary": ["enabled": false],
                "compressed": ["enabled": true, "format": "aac"]
            ]
        ] as [String : Any]
        
        let mockResult = simulateRecording(config: config)
        
        // Check main result structure
        let fileUri = mockResult["fileUri"] as? String ?? ""
        let _ = mockResult["filename"] as? String ?? ""
        let _ = mockResult["mimeType"] as? String ?? ""
        
        // Check compression structure
        let compressionDict = mockResult["compression"] as? [String: Any]
        let compressedUri = compressionDict?["compressedFileUri"] as? String
        let compressedSize = compressionDict?["size"] as? Int64
        
        // When primary is disabled, we should either:
        // 1. Get compression info with the compressed file URI
        // 2. Or use compressed URI as main fileUri (like web does)
        let hasAccessToCompressed = (compressedUri != nil && !compressedUri!.isEmpty) ||
                                   (!fileUri.isEmpty && fileUri != "")
        
        let passed = hasAccessToCompressed && compressedSize != nil
        
        results.append((
            name: "File Accessibility",
            passed: passed,
            message: "Main URI: '\(fileUri)', Compressed URI: '\(compressedUri ?? "nil")', Size: \(compressedSize ?? 0)"
        ))
        
        if passed {
            print("✅ Compressed file is accessible")
            print("✅ File size reported: \(compressedSize ?? 0) bytes")
        } else {
            print("❌ FAIL: Cannot access compressed file")
            print("   Main fileUri: '\(fileUri)'")
            print("   Compressed URI: '\(compressedUri ?? "nil")'")
        }
        print()
    }
    
    // Helper to simulate recording
    func simulateRecording(config: [String: Any]) -> [String: Any] {
        // This simulates the current BUGGY behavior
        // After the fix, this should return proper compression info
        
        let outputConfig = config["output"] as? [String: Any]
        let primaryConfig = outputConfig?["primary"] as? [String: Any]
        let compressedConfig = outputConfig?["compressed"] as? [String: Any]
        
        let primaryEnabled = primaryConfig?["enabled"] as? Bool ?? true
        let compressedEnabled = compressedConfig?["enabled"] as? Bool ?? false
        
        if !primaryEnabled {
            // Current buggy behavior - returns nil compression
            let result: [String: Any] = [
                "fileUri": "",
                "filename": "stream-only",
                "durationMs": 5000,
                "size": 0,
                "mimeType": "audio/wav"
            ]
            // BUG: compression should be included but is currently nil
            return result
        } else {
            // Normal behavior when primary is enabled
            var result: [String: Any] = [
                "fileUri": "file:///mock/recording.wav",
                "filename": "recording.wav",
                "durationMs": 5000,
                "size": 240000,
                "mimeType": "audio/wav"
            ]
            
            if compressedEnabled {
                result["compression"] = [
                    "compressedFileUri": "file:///mock/recording.aac",
                    "format": "aac",
                    "bitrate": 128000,
                    "size": 40000,
                    "mimeType": "audio/aac"
                ]
            }
            
            return result
        }
    }
    
    func printResults() {
        print("\n📊 Test Results")
        print("===============")
        
        let passed = results.filter { $0.passed }.count
        let total = results.count
        
        for result in results {
            let status = result.passed ? "✅" : "❌"
            print("\(status) \(result.name)")
            print("   \(result.message)")
        }
        
        print("\nSummary: \(passed)/\(total) tests passed")
        
        if passed == total {
            print("🎉 All tests passed!")
        } else {
            print("⚠️  Some tests failed - Fix needed!")
            print("\n🔧 Required Fix:")
            print("- When primary output is disabled, compression info must be included")
            print("- Compressed file URI must be accessible to users")
            print("- This affects iOS and Android (Web works correctly)")
        }
    }
}

// Run the test
let test = CompressedOnlyOutputTest()
test.runAllTests()